File: src/audio/dsp.py
import math
from dataclasses import dataclass, field
from typing import Optional
import numpy as np
import scipy.signal


@dataclass
class DSPOptions:
    """Configurable parameters for the real-time DSP pipeline."""

    # Master Bypass & Noise Gate
    bypass: bool = False
    noise_gate_enabled: bool = False
    noise_gate_threshold_db: float = -65.0  # dB (gentle threshold)

    # Pitch Shift
    pitch_semitones: float = 0.0  # -12 to +12 semitones

    # Robot / Ring Modulation
    robot_enabled: bool = False
    robot_freq: float = 150.0  # Hz (50Hz - 600Hz)
    robot_mix: float = 0.75

    # Radio / Walkie-Talkie
    radio_enabled: bool = False

    # Distortion / Bitcrush
    distortion_enabled: bool = False
    distortion_drive: float = 0.0  # 0.0 to 1.0

    # Reverb
    reverb_enabled: bool = False
    reverb_room_size: float = 0.0  # 0.0 to 1.0
    reverb_wet: float = 0.0  # 0.0 to 1.0

    # Chorus
    chorus_enabled: bool = False
    chorus_depth: float = 0.0  # 0.0 to 1.0
    chorus_rate: float = 1.2  # Hz

    # Filters
    highpass_cutoff: float = 20.0  # Hz
    lowpass_cutoff: float = 20000.0  # Hz

    # Master Gains
    input_gain: float = 1.0
    output_gain: float = 1.0


class ContinuousPitchShifter:
    """High-performance real-time vocal pitch shifter using vectorized 4-grain synthesis."""

    def __init__(self, sample_rate: int = 48000):
        self.sr = sample_rate
        self.buf_len = int(sample_rate * 0.25)  # 250ms circular buffer
        self.buffer = np.zeros(self.buf_len, dtype=np.float32)
        self.write_ptr = 0
        self.phases = np.array([0.0, 0.25, 0.5, 0.75], dtype=np.float32)

    def reset(self):
        self.buffer.fill(0)
        self.write_ptr = 0
        self.phases = np.array([0.0, 0.25, 0.5, 0.75], dtype=np.float32)

    def process(self, audio: np.ndarray, semitones: float) -> np.ndarray:
        if abs(semitones) < 0.05:
            return audio

        pitch_ratio = 2.0 ** (semitones / 12.0)
        rate = 1.0 - pitch_ratio

        # Adaptive grain size: larger window for deep voice to preserve pitch fundamentals,
        # smaller window for high voice for crisp definition without delay smear.
        if semitones < 0:
            window_ms = 35.0 + min(abs(semitones) * 1.5, 25.0)  # 35-60ms
        else:
            window_ms = max(20.0 - semitones * 0.6, 12.0)  # 12-20ms

        w_size = int(self.sr * (window_ms / 1000.0))
        n = len(audio)
        buf_len = self.buf_len

        # Store incoming block into circular buffer
        write_indices = (self.write_ptr + np.arange(n)) % buf_len
        self.buffer[write_indices] = audio
        curr_write_ptrs = self.write_ptr + np.arange(n)

        # Vectorized 4-grain synthesis with Hann window
        steps = np.arange(n, dtype=np.float32) * (rate / w_size)
        out = np.zeros(n, dtype=np.float32)

        for k in range(4):
            phase_k = (self.phases[k] + steps) % 1.0
            # Raised cosine window
            w = 0.5 * (1.0 - np.cos(2.0 * np.pi * phase_k))
            delay_samples = phase_k * w_size
            tap = curr_write_ptrs - delay_samples

            idx0 = np.floor(tap).astype(np.int64) % buf_len
            idx1 = (idx0 + 1) % buf_len
            frac = tap - np.floor(tap)

            grain_val = (1.0 - frac) * self.buffer[idx0] + frac * self.buffer[idx1]
            out += (w * grain_val)

        # 4 overlapping Hann windows sum to 2.0 -> normalize by 0.5
        out *= 0.5

        # Update phases
        self.phases = (self.phases + n * (rate / w_size)) % 1.0
        self.write_ptr = (self.write_ptr + n) % buf_len
        return out.astype(np.float32)


class SchroederReverb:
    """Low-latency Schroeder Reverb with 4 comb filters and 2 all-pass diffusers."""

    def __init__(self, sample_rate: int = 48000):
        self.sr = sample_rate
        # Delay lengths in samples scaled for 48kHz
        comb_delays = [
            int(sample_rate * t)
            for t in [0.0297, 0.0371, 0.0411, 0.0437]  # ~30ms to ~44ms
        ]
        allpass_delays = [
            int(sample_rate * t) for t in [0.0050, 0.0017]  # ~5ms, ~1.7ms
        ]

        self.comb_buffers = [np.zeros(d, dtype=np.float32) for d in comb_delays]
        self.comb_ptrs = [0] * len(comb_delays)

        self.ap_buffers = [np.zeros(d, dtype=np.float32) for d in allpass_delays]
        self.ap_ptrs = [0] * len(allpass_delays)

    def reset(self):
        for b in self.comb_buffers:
            b.fill(0)
        for b in self.ap_buffers:
            b.fill(0)
        self.comb_ptrs = [0] * len(self.comb_buffers)
        self.ap_ptrs = [0] * len(self.ap_buffers)

    def process(
        self, audio: np.ndarray, room_size: float, wet: float
    ) -> np.ndarray:
        if wet <= 0.001:
            return audio

        feedback = 0.65 + 0.28 * min(max(room_size, 0.0), 1.0)
        out_wet = np.zeros_like(audio)
        n = len(audio)

        # 4 Parallel Comb Filters
        for c_idx, buf in enumerate(self.comb_buffers):
            b_len = len(buf)
            ptr = self.comb_ptrs[c_idx]
            for i in range(n):
                delayed = buf[ptr]
                buf[ptr] = audio[i] + delayed * feedback
                out_wet[i] += delayed * 0.25
                ptr = (ptr + 1) % b_len
            self.comb_ptrs[c_idx] = ptr

        # 2 Series All-Pass Filters
        ap_gain = 0.5
        for ap_idx, buf in enumerate(self.ap_buffers):
            b_len = len(buf)
            ptr = self.ap_ptrs[ap_idx]
            for i in range(n):
                delayed = buf[ptr]
                in_val = out_wet[i]
                buf[ptr] = in_val + delayed * ap_gain
                out_wet[i] = delayed - in_val * ap_gain
                ptr = (ptr + 1) % b_len
            self.ap_ptrs[ap_idx] = ptr

        return audio * (1.0 - wet * 0.5) + out_wet * wet


class ChorusEffect:
    """Real-time stereo/mono chorus effect with modulated delay line."""

    def __init__(self, sample_rate: int = 48000):
        self.sr = sample_rate
        self.max_delay = int(sample_rate * 0.03)  # 30ms max delay
        self.buffer = np.zeros(self.max_delay, dtype=np.float32)
        self.write_ptr = 0
        self.lfo_phase = 0.0

    def reset(self):
        self.buffer.fill(0)
        self.write_ptr = 0
        self.lfo_phase = 0.0

    def process(
        self, audio: np.ndarray, depth: float, rate: float = 1.2
    ) -> np.ndarray:
        if depth <= 0.01:
            return audio

        n = len(audio)
        out = np.empty_like(audio)
        buf_len = self.max_delay
        lfo_step = 2.0 * math.pi * rate / self.sr
        base_delay = buf_len * 0.5
        mod_amp = buf_len * 0.35 * depth

        for i in range(n):
            self.buffer[self.write_ptr] = audio[i]
            delay = base_delay + mod_amp * math.sin(self.lfo_phase)
            self.lfo_phase += lfo_step
            if self.lfo_phase > 2.0 * math.pi:
                self.lfo_phase -= 2.0 * math.pi

            tap = self.write_ptr - delay
            idx0 = int(math.floor(tap)) % buf_len
            idx1 = (idx0 + 1) % buf_len
            frac = tap - math.floor(tap)
            delayed_val = (
                1.0 - frac
            ) * self.buffer[idx0] + frac * self.buffer[idx1]

            out[i] = audio[i] * 0.7 + delayed_val * 0.5 * depth
            self.write_ptr = (self.write_ptr + 1) % buf_len

        return out


class VoiceDSP:
    """Main Real-Time DSP Audio Processing Pipeline."""

    def __init__(self, sample_rate: int = 48000, block_size: int = 256):
        self.sr = sample_rate
        self.block_size = block_size
        self.options = DSPOptions()

        # DSP Submodules
        self.pitch_shifter = ContinuousPitchShifter(sample_rate)
        self.reverb = SchroederReverb(sample_rate)
        self.chorus = ChorusEffect(sample_rate)

        # Carrier state for Ring Modulator / Robot voice
        self.robot_carrier_phase = 0.0

        # Radio bandpass filter design (300Hz - 3400Hz)
        self._init_filters()

        # Noise Gate Envelope follower state (-100 dB represents baseline silence)
        self.noise_gate_envelope = -100.0

    def _init_filters(self):
        nyquist = self.sr * 0.5
        # Radio bandpass filter
        low = 300.0 / nyquist
        high = 3400.0 / nyquist
        self.radio_b, self.radio_a = scipy.signal.butter(
            2, [low, high], btype="bandpass"
        )
        self.radio_zi = scipy.signal.lfilter_zi(self.radio_b, self.radio_a) * 0

    def update_options(self, options: DSPOptions):
        self.options = options

    def reset(self):
        """Resets all filter and delay states."""
        self.pitch_shifter.reset()
        self.reverb.reset()
        self.chorus.reset()
        self.robot_carrier_phase = 0.0
        self.radio_zi = np.zeros_like(self.radio_zi)
        self.noise_gate_envelope = -100.0

    def process(self, input_block: np.ndarray) -> np.ndarray:
        """Processes a single 1D or 2D audio block in real-time.

        Args:
            input_block: float32 NumPy array with values normalized to [-1.0, 1.0].
        Returns:
            processed audio block with the same shape.
        """
        if self.options.bypass:
            return input_block * self.options.output_gain

        # Handle stereo input by converting or averaging for mono processing
        is_stereo = (input_block.ndim == 2 and input_block.shape[1] == 2) or (
            input_block.ndim == 2 and input_block.shape[0] == 2
        )

        if input_block.ndim == 2:
            if input_block.shape[1] == 2:
                mono = 0.5 * (input_block[:, 0] + input_block[:, 1])
            else:
                mono = 0.5 * (input_block[0, :] + input_block[1, :])
        else:
            mono = input_block.copy()

        # 1. Input Gain
        if self.options.input_gain != 1.0:
            mono = mono * self.options.input_gain

        # 2. Noise Gate (Pre-DSP) with Soft Knee
        if self.options.noise_gate_enabled:
            rms = np.sqrt(np.mean(mono**2) + 1e-12)
            db = 20.0 * np.log10(rms)
            # Fast attack (0.45) to open instantly on speech, smooth release (0.04)
            alpha = 0.45 if db > self.noise_gate_envelope else 0.04
            self.noise_gate_envelope = (
                1.0 - alpha
            ) * self.noise_gate_envelope + alpha * db

            threshold = self.options.noise_gate_threshold_db
            knee_width = 8.0  # dB
            if self.noise_gate_envelope < threshold - knee_width:
                mono *= 0.0
            elif self.noise_gate_envelope < threshold:
                gate_ratio = (
                    self.noise_gate_envelope - (threshold - knee_width)
                ) / knee_width
                mono *= float(gate_ratio)

        # 3. Pitch Shift
        if abs(self.options.pitch_semitones) >= 0.1:
            mono = self.pitch_shifter.process(
                mono, self.options.pitch_semitones
            )

        # 4. Robot Voice / Ring Modulation
        if self.options.robot_enabled and self.options.robot_freq > 0:
            n_samples = len(mono)
            step = 2.0 * math.pi * self.options.robot_freq / self.sr
            t = self.robot_carrier_phase + np.arange(n_samples) * step
            carrier = np.sin(t).astype(np.float32)
            self.robot_carrier_phase = (
                self.robot_carrier_phase + n_samples * step
            ) % (2.0 * math.pi)

            # Modulate and blend
            modulated = mono * carrier
            mix = self.options.robot_mix
            mono = mono * (1.0 - mix) + modulated * mix * 1.4

        # 5. Radio / Walkie-Talkie Filter
        if self.options.radio_enabled:
            filtered, self.radio_zi = scipy.signal.lfilter(
                self.radio_b, self.radio_a, mono, zi=self.radio_zi
            )
            # Add slight overdrive/resonance
            mono = np.tanh(filtered * 2.2) * 0.75

        # 6. Distortion / Drive
        if (
            self.options.distortion_enabled
            and self.options.distortion_drive > 0.01
        ):
            drive_factor = 1.0 + self.options.distortion_drive * 8.0
            mono = np.tanh(mono * drive_factor) * (
                1.0 / (1.0 + self.options.distortion_drive * 0.5)
            )

        # 7. Chorus
        if self.options.chorus_enabled and self.options.chorus_depth > 0.01:
            mono = self.chorus.process(
                mono, self.options.chorus_depth, self.options.chorus_rate
            )

        # 8. Reverb
        if self.options.reverb_enabled and self.options.reverb_wet > 0.01:
            mono = self.reverb.process(
                mono, self.options.reverb_room_size, self.options.reverb_wet
            )

        # 9. Output Gain & Soft Knee Limiter
        mono = mono * self.options.output_gain
        # Soft limiter using tanh to prevent harsh clipping
        over = np.abs(mono) > 0.95
        if np.any(over):
            mono = np.tanh(mono)

        # Reshape to original input format
        if input_block.ndim == 2:
            if input_block.shape[1] == 2:
                return np.column_stack((mono, mono)).astype(np.float32)
            else:
                return np.row_stack((mono, mono)).astype(np.float32)
        return mono.astype(np.float32)

File: src/src/audio/dsp.rs
use serde::{Deserialize, Serialize};
use std::f32::consts::PI;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DSPOptions {
    pub bypass: bool,
    pub noise_gate_enabled: bool,
    pub noise_gate_threshold_db: f32,

    pub pitch_semitones: f32,

    pub robot_enabled: bool,
    pub robot_freq: f32,
    pub robot_mix: f32,

    pub radio_enabled: bool,

    pub distortion_enabled: bool,
    pub distortion_drive: f32,

    pub reverb_enabled: bool,
    pub reverb_room_size: f32,
    pub reverb_wet: f32,

    pub chorus_enabled: bool,
    pub chorus_depth: f32,
    pub chorus_rate: f32,

    pub highpass_cutoff: f32,
    pub lowpass_cutoff: f32,

    pub input_gain: f32,
    pub output_gain: f32,
}

impl Default for DSPOptions {
    fn default() -> Self {
        Self {
            bypass: false,
            noise_gate_enabled: false,
            noise_gate_threshold_db: -65.0,

            pitch_semitones: 0.0,

            robot_enabled: false,
            robot_freq: 150.0,
            robot_mix: 0.75,

            radio_enabled: false,

            distortion_enabled: false,
            distortion_drive: 0.0,

            reverb_enabled: false,
            reverb_room_size: 0.0,
            reverb_wet: 0.0,

            chorus_enabled: false,
            chorus_depth: 0.0,
            chorus_rate: 1.2,

            highpass_cutoff: 20.0,
            lowpass_cutoff: 20000.0,

            input_gain: 1.0,
            output_gain: 1.0,
        }
    }
}

/// High-performance real-time vocal pitch shifter using 4-grain synthesis.
#[derive(Debug, Clone)]
pub struct ContinuousPitchShifter {
    sr: usize,
    buf_len: usize,
    buffer: Vec<f32>,
    write_ptr: usize,
    phases: [f32; 4],
}

impl ContinuousPitchShifter {
    pub fn new(sample_rate: usize) -> Self {
        let buf_len = sample_rate / 4; // 250ms circular buffer
        Self {
            sr: sample_rate,
            buf_len,
            buffer: vec![0.0; buf_len],
            write_ptr: 0,
            phases: [0.0, 0.25, 0.5, 0.75],
        }
    }

    pub fn reset(&mut self) {
        self.buffer.fill(0.0);
        self.write_ptr = 0;
        self.phases = [0.0, 0.25, 0.5, 0.75];
    }

    pub fn process(&mut self, audio: &[f32], semitones: f32, out: &mut [f32]) {
        if semitones.abs() < 0.05 {
            out.copy_from_slice(audio);
            return;
        }

        let pitch_ratio = 2.0_f32.powf(semitones / 12.0);
        let rate = 1.0 - pitch_ratio;

        let window_ms = if semitones < 0.0 {
            35.0 + (semitones.abs() * 1.5).min(25.0)
        } else {
            (20.0 - semitones * 0.6).max(12.0)
        };

        let w_size = ((self.sr as f32) * (window_ms / 1000.0)).max(16.0);
        let n = audio.len();
        let buf_len = self.buf_len;

        // Zero output buffer
        out.fill(0.0);

        for (i, &sample) in audio.iter().enumerate() {
            let curr_ptr = self.write_ptr + i;
            let write_idx = curr_ptr % buf_len;
            self.buffer[write_idx] = sample;

            let step = (i as f32) * (rate / w_size);
            let mut sum_sample = 0.0;

            for k in 0..4 {
                let phase_k = (self.phases[k] + step).rem_euclid(1.0);
                // Hann window
                let w = 0.5 * (1.0 - (2.0 * PI * phase_k).cos());
                let delay_samples = phase_k * w_size;
                let tap = (curr_ptr as f32) - delay_samples;

                let tap_floor = tap.floor();
                let idx0 = ((tap_floor as isize).rem_euclid(buf_len as isize)) as usize;
                let idx1 = (idx0 + 1) % buf_len;
                let frac = tap - tap_floor;

                let grain_val = (1.0 - frac) * self.buffer[idx0] + frac * self.buffer[idx1];
                sum_sample += w * grain_val;
            }

            out[i] = sum_sample * 0.5;
        }

        let total_step = (n as f32) * (rate / w_size);
        for k in 0..4 {
            self.phases[k] = (self.phases[k] + total_step).rem_euclid(1.0);
        }
        self.write_ptr = (self.write_ptr + n) % buf_len;
    }
}

/// Low-latency Schroeder Reverb with 4 comb filters and 2 all-pass diffusers.
#[derive(Debug, Clone)]
pub struct SchroederReverb {
    comb_buffers: Vec<Vec<f32>>,
    comb_ptrs: Vec<usize>,
    ap_buffers: Vec<Vec<f32>>,
    ap_ptrs: Vec<usize>,
    wet_temp: Vec<f32>,
}

impl SchroederReverb {
    pub fn new(sample_rate: usize) -> Self {
        let comb_times = [0.0297, 0.0371, 0.0411, 0.0437];
        let ap_times = [0.0050, 0.0017];

        let comb_buffers: Vec<Vec<f32>> = comb_times
            .iter()
            .map(|&t| vec![0.0; ((sample_rate as f64) * t).round() as usize])
            .collect();
        let comb_ptrs = vec![0; comb_buffers.len()];

        let ap_buffers: Vec<Vec<f32>> = ap_times
            .iter()
            .map(|&t| vec![0.0; ((sample_rate as f64) * t).round() as usize])
            .collect();
        let ap_ptrs = vec![0; ap_buffers.len()];

        Self {
            comb_buffers,
            comb_ptrs,
            ap_buffers,
            ap_ptrs,
            wet_temp: Vec::with_capacity(1024),
        }
    }

    pub fn reset(&mut self) {
        for b in &mut self.comb_buffers {
            b.fill(0.0);
        }
        for b in &mut self.ap_buffers {
            b.fill(0.0);
        }
        self.comb_ptrs.fill(0);
        self.ap_ptrs.fill(0);
    }

    pub fn process(&mut self, audio: &[f32], room_size: f32, wet: f32, out: &mut [f32]) {
        if wet <= 0.001 {
            out.copy_from_slice(audio);
            return;
        }

        let n = audio.len();
        if self.wet_temp.len() < n {
            self.wet_temp.resize(n, 0.0);
        }
        self.wet_temp[..n].fill(0.0);

        let feedback = 0.65 + 0.28 * room_size.clamp(0.0, 1.0);

        // 4 Parallel Comb Filters
        for (c_idx, buf) in self.comb_buffers.iter_mut().enumerate() {
            let b_len = buf.len();
            let mut ptr = self.comb_ptrs[c_idx];
            for (i, &sample) in audio.iter().enumerate().take(n) {
                let delayed = buf[ptr];
                buf[ptr] = sample + delayed * feedback;
                self.wet_temp[i] += delayed * 0.25;
                ptr = (ptr + 1) % b_len;
            }
            self.comb_ptrs[c_idx] = ptr;
        }

        // 2 Series All-Pass Filters
        let ap_gain = 0.5;
        for (ap_idx, buf) in self.ap_buffers.iter_mut().enumerate() {
            let b_len = buf.len();
            let mut ptr = self.ap_ptrs[ap_idx];
            for i in 0..n {
                let delayed = buf[ptr];
                let in_val = self.wet_temp[i];
                buf[ptr] = in_val + delayed * ap_gain;
                self.wet_temp[i] = delayed - in_val * ap_gain;
                ptr = (ptr + 1) % b_len;
            }
            self.ap_ptrs[ap_idx] = ptr;
        }

        for i in 0..n {
            out[i] = audio[i] * (1.0 - wet * 0.5) + self.wet_temp[i] * wet;
        }
    }
}

/// Real-time chorus effect with modulated delay line.
#[derive(Debug, Clone)]
pub struct ChorusEffect {
    sr: usize,
    max_delay: usize,
    buffer: Vec<f32>,
    write_ptr: usize,
    lfo_phase: f32,
}

impl ChorusEffect {
    pub fn new(sample_rate: usize) -> Self {
        let max_delay = (sample_rate as f32 * 0.03) as usize; // 30ms max delay
        Self {
            sr: sample_rate,
            max_delay,
            buffer: vec![0.0; max_delay],
            write_ptr: 0,
            lfo_phase: 0.0,
        }
    }

    pub fn reset(&mut self) {
        self.buffer.fill(0.0);
        self.write_ptr = 0;
        self.lfo_phase = 0.0;
    }

    pub fn process(&mut self, audio: &[f32], depth: f32, rate: f32, out: &mut [f32]) {
        if depth <= 0.01 {
            out.copy_from_slice(audio);
            return;
        }

        let n = audio.len();
        let buf_len = self.max_delay;
        let lfo_step = 2.0 * PI * rate / (self.sr as f32);
        let base_delay = (buf_len as f32) * 0.5;
        let mod_amp = (buf_len as f32) * 0.35 * depth;

        for i in 0..n {
            self.buffer[self.write_ptr] = audio[i];
            let delay = base_delay + mod_amp * self.lfo_phase.sin();
            self.lfo_phase += lfo_step;
            if self.lfo_phase > 2.0 * PI {
                self.lfo_phase -= 2.0 * PI;
            }

            let tap = (self.write_ptr as f32) - delay;
            let tap_floor = tap.floor();
            let idx0 = ((tap_floor as isize).rem_euclid(buf_len as isize)) as usize;
            let idx1 = (idx0 + 1) % buf_len;
            let frac = tap - tap_floor;

            let delayed_val = (1.0 - frac) * self.buffer[idx0] + frac * self.buffer[idx1];
            out[i] = audio[i] * 0.7 + delayed_val * 0.5 * depth;
            self.write_ptr = (self.write_ptr + 1) % buf_len;
        }
    }
}

/// 2nd-order IIR Biquad filter (Direct Form II Transposed).
#[derive(Debug, Clone)]
pub struct BiquadFilter {
    b0: f32,
    b1: f32,
    b2: f32,
    a1: f32,
    a2: f32,
    s1: f32,
    s2: f32,
}

impl BiquadFilter {
    /// Telephone bandpass (300 Hz – 3400 Hz) matching Python's
    /// `scipy.signal.butter(2, [300, 3400], btype="bandpass")` target
    /// response: flat 0 dB passband with steep skirts.
    ///
    /// Uses the standard RBJ "band pass (constant 0 dB peak gain)" design
    /// with Q = f0 / BW, where f0 is the geometric center and BW the
    /// bandwidth in Hz.
    pub fn new_butterworth_bandpass(sample_rate: f32, f_low: f32, f_high: f32) -> Self {
        let f0 = (f_low * f_high).sqrt();
        let q = f0 / (f_high - f_low).max(1.0);
        let w0 = 2.0 * PI * f0 / sample_rate;
        let alpha = w0.sin() / (2.0 * q.max(0.05));

        let b0 = alpha;
        let b1 = 0.0;
        let b2 = -alpha;
        let a0 = 1.0 + alpha;
        let a1 = -2.0 * w0.cos();
        let a2 = 1.0 - alpha;

        Self {
            b0: b0 / a0,
            b1: b1 / a0,
            b2: b2 / a0,
            a1: a1 / a0,
            a2: a2 / a0,
            s1: 0.0,
            s2: 0.0,
        }
    }

    pub fn reset(&mut self) {
        self.s1 = 0.0;
        self.s2 = 0.0;
    }

    pub fn process_sample(&mut self, x: f32) -> f32 {
        let y = self.b0 * x + self.s1;
        self.s1 = self.b1 * x - self.a1 * y + self.s2;
        self.s2 = self.b2 * x - self.a2 * y;
        y
    }
}

/// Complete Vocal DSP processing engine.
#[derive(Debug, Clone)]
pub struct VoiceDSP {
    pub sample_rate: usize,
    pub options: DSPOptions,

    pitch_shifter: ContinuousPitchShifter,
    reverb: SchroederReverb,
    chorus: ChorusEffect,
    radio_filter: BiquadFilter,

    robot_carrier_phase: f32,
    noise_gate_envelope: f32,

    scratch_a: Vec<f32>,
    scratch_b: Vec<f32>,
}

impl VoiceDSP {
    pub fn new(sample_rate: usize, block_size: usize) -> Self {
        let radio_filter =
            BiquadFilter::new_butterworth_bandpass(sample_rate as f32, 300.0, 3400.0);

        Self {
            sample_rate,
            options: DSPOptions::default(),

            pitch_shifter: ContinuousPitchShifter::new(sample_rate),
            reverb: SchroederReverb::new(sample_rate),
            chorus: ChorusEffect::new(sample_rate),
            radio_filter,

            robot_carrier_phase: 0.0,
            noise_gate_envelope: -100.0,

            scratch_a: vec![0.0; block_size.max(1024)],
            scratch_b: vec![0.0; block_size.max(1024)],
        }
    }

    pub fn update_options(&mut self, options: DSPOptions) {
        self.options = options;
    }

    pub fn reset(&mut self) {
        self.pitch_shifter.reset();
        self.reverb.reset();
        self.chorus.reset();
        self.radio_filter.reset();
        self.robot_carrier_phase = 0.0;
        self.noise_gate_envelope = -100.0;
    }

    /// Process mono block of audio samples in place or into destination buffer.
    pub fn process(&mut self, input: &[f32], output: &mut [f32]) {
        let n = input.len();
        if output.len() < n {
            return;
        }

        if self.scratch_a.len() < n {
            self.scratch_a.resize(n, 0.0);
            self.scratch_b.resize(n, 0.0);
        }

        if self.options.bypass {
            let gain = self.options.output_gain;
            for (i, &sample) in input.iter().enumerate().take(n) {
                output[i] = sample * gain;
            }
            return;
        }

        // 1. Input Gain
        let in_gain = self.options.input_gain;
        for (i, &sample) in input.iter().enumerate().take(n) {
            self.scratch_a[i] = sample * in_gain;
        }

        // 2. Noise Gate (Soft Knee)
        if self.options.noise_gate_enabled {
            let mut sum_sq = 0.0;
            for &sample in self.scratch_a.iter().take(n) {
                sum_sq += sample * sample;
            }
            let rms = (sum_sq / (n as f32) + 1e-12).sqrt();
            let db = 20.0 * rms.log10();

            let alpha = if db > self.noise_gate_envelope {
                0.45
            } else {
                0.04
            };
            self.noise_gate_envelope = (1.0 - alpha) * self.noise_gate_envelope + alpha * db;

            let threshold = self.options.noise_gate_threshold_db;
            let knee_width = 8.0;

            if self.noise_gate_envelope < threshold - knee_width {
                self.scratch_a[..n].fill(0.0);
            } else if self.noise_gate_envelope < threshold {
                let gate_ratio =
                    (self.noise_gate_envelope - (threshold - knee_width)) / knee_width;
                for sample in self.scratch_a.iter_mut().take(n) {
                    *sample *= gate_ratio;
                }
            }
        }

        // 3. Pitch Shift
        if self.options.pitch_semitones.abs() >= 0.1 {
            self.pitch_shifter.process(
                &self.scratch_a[..n],
                self.options.pitch_semitones,
                &mut self.scratch_b[..n],
            );
            self.scratch_a[..n].copy_from_slice(&self.scratch_b[..n]);
        }

        // 4. Robot Voice / Ring Modulation
        if self.options.robot_enabled && self.options.robot_freq > 0.0 {
            let step = 2.0 * PI * self.options.robot_freq / (self.sample_rate as f32);
            let mix = self.options.robot_mix;
            for i in 0..n {
                let carrier = (self.robot_carrier_phase + (i as f32) * step).sin();
                let modulated = self.scratch_a[i] * carrier;
                self.scratch_a[i] = self.scratch_a[i] * (1.0 - mix) + modulated * mix * 1.4;
            }
            self.robot_carrier_phase =
                (self.robot_carrier_phase + (n as f32) * step).rem_euclid(2.0 * PI);
        }

        // 5. Radio / Walkie-Talkie Filter
        if self.options.radio_enabled {
            for sample in self.scratch_a.iter_mut().take(n) {
                let filtered = self.radio_filter.process_sample(*sample);
                *sample = (filtered * 2.2).tanh() * 0.75;
            }
        }

        // 6. Distortion / Drive
        if self.options.distortion_enabled && self.options.distortion_drive > 0.01 {
            let drive_factor = 1.0 + self.options.distortion_drive * 8.0;
            let norm = 1.0 / (1.0 + self.options.distortion_drive * 0.5);
            for sample in self.scratch_a.iter_mut().take(n) {
                *sample = (*sample * drive_factor).tanh() * norm;
            }
        }

        // 7. Chorus
        if self.options.chorus_enabled && self.options.chorus_depth > 0.01 {
            self.chorus.process(
                &self.scratch_a[..n],
                self.options.chorus_depth,
                self.options.chorus_rate,
                &mut self.scratch_b[..n],
            );
            self.scratch_a[..n].copy_from_slice(&self.scratch_b[..n]);
        }

        // 8. Reverb
        if self.options.reverb_enabled && self.options.reverb_wet > 0.01 {
            self.reverb.process(
                &self.scratch_a[..n],
                self.options.reverb_room_size,
                self.options.reverb_wet,
                &mut self.scratch_b[..n],
            );
            self.scratch_a[..n].copy_from_slice(&self.scratch_b[..n]);
        }

        // 9. Output Gain & Soft Limiter
        let out_gain = self.options.output_gain;
        for (i, out_sample) in output.iter_mut().enumerate().take(n) {
            let mut val = self.scratch_a[i] * out_gain;
            if val.abs() > 0.95 {
                val = val.tanh();
            }
            *out_sample = val;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dsp_bypass() {
        let mut dsp = VoiceDSP::new(48000, 256);
        dsp.options.bypass = true;
        let input = vec![0.5; 256];
        let mut output = vec![0.0; 256];
        dsp.process(&input, &mut output);
        assert_eq!(input, output);
    }

    #[test]
    fn test_pitch_shifter() {
        let mut dsp = VoiceDSP::new(48000, 256);
        dsp.options.pitch_semitones = 4.0;
        let input = vec![0.2; 256];
        let mut output = vec![0.0; 256];
        dsp.process(&input, &mut output);
        // On subsequent blocks grain synthesis outputs shifted signals
        dsp.process(&input, &mut output);
        assert!(output.iter().any(|&x| x.abs() > 0.001));
        assert!(output.iter().all(|&x| x.is_finite()));
    }

    #[test]
    fn test_reverb_and_chorus() {
        let mut dsp = VoiceDSP::new(48000, 256);
        dsp.options.reverb_enabled = true;
        dsp.options.reverb_wet = 0.5;
        dsp.options.chorus_enabled = true;
        dsp.options.chorus_depth = 0.5;
        let input = vec![0.1; 256];
        let mut output = vec![0.0; 256];
        dsp.process(&input, &mut output);
        assert!(output.iter().all(|&x| x.is_finite()));
    }
}
